core/domain/job_services_test.py

```diff
@@ -0,0 +1,332 @@
+                job_services.get_next_job()
+
+    def test_job_validation_raises_exception_on_invalid_algorithm_id(self):
+        """Test that job validation raises exception if algorithm_id is invalid
+        in the received job data.
+        """
+        job_data = {
+            'job_id': '1',
+            'algorithm_id': 123,
+            'training_data': [],
+            'algorithm_version': 1
+        }
+        post_callback = self._get_post_callback_for_next_job(job_data)
+
+        with self.set_job_request_post_callback(post_callback):
+            with self.assertRaisesRegexp(
+                Exception, 'Expected algorithm id to be unicode'):
+                job_services.get_next_job()
+
+    def test_job_validation_raises_exception_on_invalid_training_data(self):
+        """Test that job validation raises exception if training_data is invalid
+        in the received job data.
+        """
+        job_data = {
+            'job_id': '1',
+            'algorithm_id': 'TextClassifier',
+            'training_data': 'data',
+            'algorithm_version': 1
+        }
+        post_callback = self._get_post_callback_for_next_job(job_data)
+
+        with self.set_job_request_post_callback(post_callback):
+            with self.assertRaisesRegexp(
+                Exception, 'Expected training data to be a list'):
+                job_services.get_next_job()
+
+        job_data['training_data'] = [{'answers': ['a', 'b', 'c']}]
+        post_callback = self._get_post_callback_for_next_job(job_data)
+
+        with self.set_job_request_post_callback(post_callback):
+            with self.assertRaisesRegexp(
+                Exception,
+                'Expected answer_group_index to be a key in training_data'):
+                job_services.get_next_job()
+
+        job_data['training_data'] = [{'answer_group_index': 1}]
+        post_callback = self._get_post_callback_for_next_job(job_data)
+
+        with self.set_job_request_post_callback(post_callback):
+            with self.assertRaisesRegexp(
+                Exception,
+                'Expected answers to be a key in training_data'):
+                job_services.get_next_job()
+
+        job_data['training_data'] = [{
+            'answer_group_index': '1',
+            'answers': ['a', 'b', 'c']
+        }]
+        post_callback = self._get_post_callback_for_next_job(job_data)
+
+        with self.set_job_request_post_callback(post_callback):
+            with self.assertRaisesRegexp(
+                Exception,
+                'Expected answer_group_index to be an int'):
+                job_services.get_next_job()
+
+        job_data['training_data'] = [{
+            'answer_group_index': 1,
+            'answers': 'answer1'
+        }]
+        post_callback = self._get_post_callback_for_next_job(job_data)
+        with self.set_job_request_post_callback(post_callback):
+            with self.assertRaisesRegexp(
```

Comment:
Suggest splitting up the tests in here, as well. This test seems to be verifying a bunch of different behaviors--these would be a bit cleaner if in separate tests.

Reply:
Done